be/src/exec/pipeline/query_context.cpp

```diff
@@ -96,6 +96,14 @@ Status QueryContext::init_query(workgroup::WorkGroup* wg) {
     return st;
 }
 
+void QueryContext::set_query_trace(std::shared_ptr<starrocks::debug::QueryTrace> query_trace) {
+    bool expected = false;
+    // make sure _query_trace is initilized only once
+    if (_is_query_trace_inited.compare_exchange_strong(expected, true)) {
+        _query_trace = std::move(query_trace);
+    }
```

Comment:
Maybe be use `std::call_once`? Otherwise, is that possible?
1. Thread#1 set `_is_query_trace_inited` as true.
2. Before Thread#1 set `_query_trace`, Thread#2 uses _query_trace and may cause NPE.

Reply:
> Maybe be use `std::call_once`? Otherwise, is that possible?
> 
> 1. Thread#1 set `_is_query_trace_inited` as true.
> 2. Before Thread#1 set `_query_trace`, Thread#2 uses `_query_trace`, which may cause NPE.

I think this situation can not occur. `set_query_trace` is only called in fragment prepare phase, nowhere accesses it at the same time.

Reply:
It is possible that two fragments of a query are paraparing and calling  `set_query_trace` concurrently?

Reply:
> Is it possible that two fragments of a query are paraparing and calling `set_query_trace` concurrently?

I think it may happen, but use this cas operation, only one can set successfully.